### PR TITLE
Improve logging and block direct requests to server IP

### PR DESCRIPTION
**Pull request by BenHarris**
- Tidied up logs
- Trimmed down request latency
- Simplified colours
- Implemented blocking of requests to the server's IP (during testing a spambot started hammering me!) - using http://checkip.amazonaws.com at the moment, but could use another service or just use this as a fallback if it is not supplied as a parameter?

### CODE REVIEW

**Comment by antoniomika** [comment]
Hey @BenHarris! Thanks for the PR! Maybe instead of calling on the external, we can just do a check to see if the request was made to an IP address and fail out? Also, mind adding some documentation to the round functions?

**Comment by BenHarris** [comment]
Sure. I’m not 100% happy with the rounding stuff. Pretty sure there must be a simpler way. Will see what I can do 😀.

I had a few other thoughts about using the external IP, but can switch it to just detect if it’s an IP for now.

**Comment by BenHarris** [comment]
Just realised I wasn't getting Github emails 🤦‍♂. Have updated that now. Also implemented an initial version of the log detail functionality. It might be worth always logging the full version to the server, but it's a handy option for now.

**Comment by antoniomika** [comment]
Sorry for all of the noise! Looks like CI is setup properly now! Would you mind taking care of those issues and then we can looking at getting things moving?

**Comment by BenHarris** [comment]
I'm working on a much more flexible implementation of this now

**Comment by BenHarris** [comment]
@antoniomika let me know what you think!

Realised the log format wasn't that useful when it was the same for client and server, so created two log formats for client and server.

**Comment by antoniomika** [comment]
@BenHarris Awesome! I'll do a review on it real quick!

File: README.md [hunk 1]
```diff
@@ -154,6 +156,10 @@ Usage of ./sish:
         Whether or not to redirect the root domain (default true)
   -sish.redirectrootlocation string
         Where to redirect the root domain to (default "https://github.com/antoniomika/sish")
+  -sish.serverlogformat string
```
Comment (antoniomika) [inline README.md]: This help output is generated from the data stored as `flag` comments. Would you mind updated those to reflect these and then regenerating this help output from a `go run . -h`?

**Comment by BenHarris** [comment]
There was some weirdness when using newlines directly in the log definition, hence using {newline}

File: http.go [hunk 2]
```diff
@@ -161,3 +155,115 @@ func startHTTPHandler(state *State) {
 	}
 	log.Fatal(r.Run(*httpAddr))
 }
+
+// Round a duration, see original source here: https://play.golang.org/p/WjfKwhhjL5
+func Round(d, r time.Duration) time.Duration {
```
Comment (antoniomika) [inline http.go]: Does the [time().Round()](https://golang.org/pkg/time/#Time.Round) function not do what you want here?

Comment (BenHarris) [reply http.go]: I don't think so. I was originally trying to do it with time().Round(), but was unable to get the fractional output I was aiming for

Comment (BenHarris) [reply http.go]: Take it back it does seem to work

File: README.md (hunk 1 above)
Comment (BenHarris) [reply README.md]: Gotcha, no problem

File: http.go [hunk 3]
```diff
@@ -41,41 +42,34 @@ func startHTTPHandler(state *State) {
...
-		)
-
+		// Client log
 		if *logToClient {
 			hostname := strings.Split(param.Request.Host, ":")[0]
 			loc, ok := state.HTTPListeners.Load(hostname)
 			if ok {
 				proxyHolder := loc.(*ProxyHolder)
-				sendMessage(proxyHolder.SSHConn, strings.TrimSpace(logLine), true)
+				sendMessage(proxyHolder.SSHConn, strings.TrimSpace(formatLog(clientLogFormatParts, &param)), true)
```
Comment (antoniomika) [inline http.go]: Let's store this as a variable so we only have to string process it once

File: http.go [hunk 4]
```diff
@@ -41,41 +42,34 @@ func startHTTPHandler(state *State) {
...
 			}
 		}
 
-		return logLine
+		// Server log line
+		return formatLog(serverLogFormatParts, &param) + "\n"
```
Comment (antoniomika) [inline http.go]: Move the newline to the `formatLog` method. That's why TrimSpace is used when sending the message to the channel.

File: http.go [hunk 5]
```diff
@@ -161,3 +155,96 @@ func startHTTPHandler(state *State) {
...
+		d /= 10
+		i++
+	}
+	return i
+}
+
+func formatLog(logFormatParts []string, param *gin.LogFormatterParams) string {
+	statusFormatted := fmt.Sprintf("%3d", param.StatusCode)
+
+	if param.IsOutputColor() {
```
Comment (antoniomika) [inline http.go]: Let's also add the color back for the method. That was useful for quickly seeing request types as they came in. I think I also prefer having the background color as the default, but allowing for it to be changed would be nice too.

File: requests.go [hunk 6]
```diff
@@ -107,7 +107,7 @@ func handleRemoteForward(newRequest *ssh.Request, sshConn *SSHConnection, state
 		connType = "https"
 	}
 
-	requestMessages := fmt.Sprintf("Starting SSH Fowarding service for %s. Forwarded connections can be accessed via the following methods:\r\n", aurora.Sprintf(aurora.Green("%s:%s"), connType, stringPort))
+	requestMessages := fmt.Sprintf("Forwarding %s traffic on port %s from:\r\n", connType, stringPort)
```
Comment (antoniomika) [inline requests.go]: I think this needs to be a bit more descriptive

**Comment by antoniomika** [comment]
Closing for now as there have been quite a few changes in the v1 branch that likely solve many of these issues

**Pull request closed without merging**